sqlserver/datadog_checks/sqlserver/stored_procedures.py

```diff
@@ -0,0 +1,186 @@
+        cursor.execute("select top 0 * from sys.dm_exec_procedure_stats")
+        all_columns = {i[0] for i in cursor.description}
+        available_columns = [c for c in all_expected_columns if c in all_columns]
+        missing_columns = set(all_expected_columns) - set(available_columns)
+        if missing_columns:
+            self.log.debug(
+                "missing the following expected query metrics columns from dm_exec_procedure_stats: %s", missing_columns
+            )
+        self.log.debug("found available sys.dm_exec_procedure_stats columns: %s", available_columns)
+        return available_columns
+
+    def _get_procedure_metrics_query_cached(self, cursor):
+        if self._procedure_metrics_query:
+            return self._procedure_metrics_query
+        available_columns = self._get_available_procedure_metrics_columns(cursor, SQL_SERVER_PROCEDURE_METRICS_COLUMNS)
+        self._procedure_metrics_query = PROCEDURE_METRICS_QUERY.format(
+            procedure_metrics_columns=', '.join(['sum({}) as {}'.format(c, c) for c in available_columns]),
+            limit=self.dm_exec_procedure_stats_row_limit,
+        )
+        return self._procedure_metrics_query
+
+    @tracked_method(agent_check_getter=agent_check_getter, track_result_length=True)
+    def _load_raw_procedure_metrics_rows(self, cursor):
+        self.log.debug("collecting sql server procedure metrics")
+        procedure_metrics_query = self._get_procedure_metrics_query_cached(cursor)
+        now = time.time()
+        self._last_stats_query_time = now
+        self.log.debug("Running query [%s]", procedure_metrics_query)
+        cursor.execute(procedure_metrics_query)
+        columns = [i[0] for i in cursor.description]
+        # construct row dicts manually as there's no DictCursor for pyodbc
+        rows = [dict(zip(columns, row)) for row in cursor.fetchall()]
+        return rows
+
+    def _collect_metrics_rows(self, cursor):
+        rows = self._load_raw_procedure_metrics_rows(cursor)
+        rows = [r for r in rows if r.get('procedure_name') and r.get('database_name')]
+        if not rows:
+            return []
+        metric_columns = [c for c in rows[0].keys() if c.startswith("total_") or c == 'execution_count']
+        rows = self._state.compute_derivative_rows(rows, metric_columns, key=_row_key)
+        return rows
+
+    def _to_metrics_payload(self, rows, max_queries):
+        # sort by total_elapsed_time and return the top max_queries
+        rows = sorted(rows, key=lambda i: i['total_elapsed_time'], reverse=True)
+        rows = rows[:max_queries]
+        return {
+            'host': self.check.resolved_hostname,
+            'timestamp': time.time() * 1000,
+            'min_collection_interval': self.collection_interval,
+            'tags': self.tags,
+            'cloud_metadata': self.check.cloud_metadata,
+            'collection_type': 'procedure_metrics',
```

Comment:
@alexandre-normand do you think we need this? I think we can just add these metrics to current sqlserver query metrics code. Is there a reason to differentiate between payloads at all?

Reply:
The rows are similar but not exactly the same. I think we could make it work without this type field but we'd essentially have to fork code paths at some point based on some other piece of info (like `procedure_name`) that would indicate it's procedure metrics. It might be a little more clear the way Joel has it set up but I might change my mind when we dive into it tomorrow (our goal is to figure out what work needs to be done to support this in the backend). 

That said, I'm not sure about the name of the field. _collection_ doesn't add much. `metric_type` might be confusing. `dbm_type` is used on events but it's not very descriptive. Maybe `type` is enough since we're already in the context of metrics payloads.

Reply:
Since the function is called `to_metrics_payload` I could call it `payload_type`

Reply:
`payload_type` works for me.